sql/src/main/java/io/crate/planner/Planner.java

```diff
@@ -377,5 +407,63 @@ private static LegacyUpsertById processInsertStatement(InsertFromValuesAnalyzedS
     public Functions functions() {
         return functions;
     }
+
+    private boolean allowSql(final AnalyzedStatement analyzedStatement) {
+        return hasValidLicense.getAsBoolean()
+               || analyzedStatement instanceof SetLicenseAnalyzedStatement
+               || analyzedStatement instanceof QueriedRelation;
+    }
+
+    private final class AllowQueryContext {
+
+        private final boolean hasValidLicense;
+
+        AllowQueryContext(boolean hasValidLicense) {
+            this.hasValidLicense = hasValidLicense;
+        }
+
+        boolean hasValidLicense() {
+            return hasValidLicense;
+        }
+    }
+
+    private static final class AllowQueryRelationVisitor extends AnalyzedRelationVisitor<AllowQueryContext, Boolean> {
```

Comment:
Including `Visitor` in the name is imho a bit of an anti-pattern if the `Visitor` aspect is not relevant to what the class is doing. It's an implementation detail in this case. 

See also my previous comment. I think we could limit the responsibility of this to answer the question "is this a read query on system tables". The name can than reflect that.